Source/WebCore/page/LocalFrame.cpp

```diff
@@ -1381,6 +1383,59 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
         view->setCanHaveScrollbars(m_scrollingMode != ScrollbarMode::AlwaysOff);
 }
 
+#if ENABLE(CONTENT_EXTENSIONS)
+
+ResourceMonitor* LocalFrame::resourceMonitor()
+{
+    constexpr bool forceEnableResourceMonitoring = true;
+    if (!isMainFrame() && (settings().iFrameResourceMonitoringEnabled() || forceEnableResourceMonitoring) && !m_resourceMonitor)
+        m_resourceMonitor = ResourceMonitor::create(*this);
+
+    return m_resourceMonitor.get();
+}
+
+ResourceMonitor* LocalFrame::parentResourceMonitor()
+{
+    if (isMainFrame())
+        return nullptr;
+
+    if (RefPtr parentFrame = dynamicDowncast<LocalFrame>(tree().parent()))
+        return parentFrame->resourceMonitor();
+
+    return nullptr;
+}
+
+void LocalFrame::networkUsageDidExceedThreshold()
+{
+    ASSERT(!isMainFrame());
+
+    // If the frame has sticky user activation, don't do offloading.
+    if (RefPtr protectedWindow = window(); protectedWindow && protectedWindow->hasStickyActivation())
+        return;
+
+    showResourceMonitoringError("This frame was unloaded."_s);
+
+    if (RefPtr document = this->document())
+        document->addConsoleMessage(MessageSource::ContentBlocker, MessageLevel::Error, "Frame was unloaded because its network usage exceeded the limit."_s);
+}
+
+void LocalFrame::showResourceMonitoringError(String&& htmlContent)
+{
+    ASSERT(!isMainFrame());
+
+    for (RefPtr<Frame> frame = this; frame; frame = frame->tree().traverseNext()) {
+        if (RefPtr localFrame = dynamicDowncast<LocalFrame>(frame)) {
+            if (RefPtr window = localFrame->window())
+                window->removeAllEventListeners();
```

Comment:
I don't think we should do this if the `RefPtr iframeElement = dynamicDowncast<HTMLIFrameElement>(ownerElement())` check fails later on. We'd break content without offloading it.